### PR TITLE
508 [COGNITION, SCREENREADER] Add descriptive labels to file upload buttons & fields

### PR DESCRIPTION
**Pull request by Mottie**
## Description

File upload buttons & fields should be better associated with the file name to which they apply, e.g. screenreaders should read the file name when targeting the delete button or any form elements inside the card.

## Original issue(s)

https://github.com/department-of-veterans-affairs/va.gov-team/issues/19696

## Testing done

Added unit tests

## Screenshots

<details><summary>Added <code>aria-describedby</code> to form elements</summary>

Screen Shot 2021-10-01 at 11 59 28 AM</details>

## Acceptance criteria
- [x] Labels & buttons have descriptive labels
- [x] All test passing

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/platform/forms-system/src/js/widgets/SelectWidget.jsx [hunk 1]
```diff
@@ -28,14 +28,20 @@ function SelectWidget({
...
+
   return (
     <select
       id={id}
       name={id}
       multiple={multiple}
-      className={`${options.widgetClassNames} ${
-        disabled ? 'vads-u-background-color--gray-lighter' : ''
+      className={`${options.widgetClassNames || ''}${
+        disabled ? ' vads-u-background-color--gray-lighter' : ''
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/widgets/SelectWidget.jsx:44]: Is this related to the `aria-describedby` changes?

File: src/platform/forms-system/src/js/widgets/SelectWidget.jsx [hunk 2]
```diff
@@ -51,6 +57,8 @@ function SelectWidget({
         const newValue = getValue(event, multiple);
         onChange(processValue(schema, newValue));
       }}
+      {...widgetProps}
+      {...(value && selectedProps[value]) || {}}
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/widgets/SelectWidget.jsx:61]: What is this doing?

File: src/platform/forms-system/src/js/fields/FileField.jsx [hunk 3]
```diff
@@ -334,6 +334,22 @@ class FileField extends React.Component {
               const deleteButtonText =
                 enableShortWorkflow && hasErrors ? 'Cancel' : 'Delete file';
 
+              const fileId = `file-name-${index}`;
+
+              const getUiSchema = innerUiSchema =>
+                typeof innerUiSchema === 'function'
+                  ? innerUiSchema({ fileId, index })
+                  : innerUiSchema;
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:342]: How does this relate to the labeling changes?

File: src/platform/forms-system/test/js/widgets/SelectWidget.unit.spec.jsx [hunk 4]
```diff
@@ -10,7 +10,7 @@ describe('Schemaform <SelectWidget>', () => {
     const enumOptions = [
       {
         label: 'Testing',
-        val: 'test',
+        value: 'testValue',
```
Comment (bkjohnson) [inline src/platform/forms-system/test/js/widgets/SelectWidget.unit.spec.jsx:13]: Was this just a bug in the test?

File: src/platform/forms-system/src/js/fields/FileField.jsx [hunk 5]
```diff
@@ -334,6 +334,22 @@ class FileField extends React.Component {
...
+
+              // make index available to widgets in attachment ui schema
+              const indexedRegistry = {
+                ...registry,
+                formContext: {
+                  ...registry.formContext,
+                  pagePerItemIndex: index,
```
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:349]: I see that we're testing for `pagePerItemIndex` being there, but what does this do for us? `Ctrl + F` only finds 3 matches for it in the PR - one here and two in the tests.

File: src/platform/forms-system/src/js/widgets/SelectWidget.jsx (hunk 1 above)
Comment (Mottie) [reply src/platform/forms-system/src/js/widgets/SelectWidget.jsx:44]: No, but it was added an "undefined" class name to the select.

File: src/platform/forms-system/src/js/widgets/SelectWidget.jsx (hunk 2 above)
Comment (Mottie) [reply src/platform/forms-system/src/js/widgets/SelectWidget.jsx:61]: It's similar to what the radio widget is doing, showing a dynamic prop when a specific item is selected. If an `aria-describedby` points to an ID that doesn't exist, but shows up when a value is selected, it will cause an axe error regarding an ID not found in the DOM.

File: src/platform/forms-system/src/js/fields/FileField.jsx (hunk 3 above)
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:342]: The file upload may or may not include an `attachmentSchema` or `attachmentName` in the `uiSchema`. And since the `aria-describedby` needs to be in the attached to whatever form element is included it gets defined in the `widgetProps` value of the `ui:options`.

The `fileId` needs to be passed to the uiSchema (code below copied from the all-claims changes in this PR) which can only be done if the `attachmentSchema` is a function.

```js
attachmentSchema: ({ fileId }) => ({
  'ui:title': 'Document type',
  'ui:disabled': isDisabled,
  'ui:widget': widgetType,
  'ui:options': {
    widgetProps: {
      'aria-describedby': fileId,
    },
  },
}),
```

File: src/platform/forms-system/src/js/fields/FileField.jsx (hunk 5 above)
Comment (Mottie) [reply src/platform/forms-system/src/js/fields/FileField.jsx:349]: Some file uploads include a `attachmentSchema` or `attachmentName` which may include a component that checks for a `pagerPerItemIndex` which is undefined in this situation.

File: src/platform/forms-system/test/js/widgets/SelectWidget.unit.spec.jsx (hunk 4 above)
Comment (Mottie) [reply src/platform/forms-system/test/js/widgets/SelectWidget.unit.spec.jsx:13]: I think so... it should be `value` everywhere, but since all the tests were passing, I changed the actual value to `testValue` to make sure.

**Review by ashagross — APPROVED** [review]

**Review by bkjohnson — DISMISSED** [review]

**Review by bkjohnson — COMMENTED** [review]
@joshkimux before I approve I just want to make sure this solves the problems in your issue.

**Review by bkjohnson — APPROVED** [review]
